packages/router/src/router.ts

```diff
@@ -636,6 +636,12 @@ export class Router {
                            (this.onSameUrlNavigation === 'reload' ? true : urlTransition) &&
                            this.urlHandlingStrategy.shouldProcessUrl(t.rawUrl);
 
+                       // If the source of the navigation is from a browser event, the URL is
+                       // already updated. We already need to sync the internal state.
+                       if (t.source !== 'imperative') {
```

Comment:
Thanks for adding comments, they help a lot with review 👍 

Small proposal (feel free to ignore): I was thinking that we can have a small helper function to make the code more self-documented:
```suggestion
                       if (isBrowserTriggeredNavigation(t.source)) {
```
Such function can be used in a couple places in this PR and also in other parts of the codebase.